client/proxmark3.c

```diff
@@ -70,16 +71,63 @@ byte_t* prx = rx;
 static void *uart_receiver(void *targ) {
 	struct receiver_arg *arg = (struct receiver_arg*)targ;
 	size_t rxlen;
+	bool need_reconnect = false;
+	UsbCommand version_cmd = {CMD_VERSION};
+	static bool request_version = false;
 
 	while (arg->run) {
-		rxlen = 0;
-		if (uart_receive(sp, prx, sizeof(UsbCommand) - (prx-rx), &rxlen) && rxlen) {
-			prx += rxlen;
-			if (prx-rx < sizeof(UsbCommand)) {
+		if( need_reconnect ) {
+			sp = uart_open(sp_name);
+
+			if( sp == INVALID_SERIAL_PORT || sp == CLAIMED_SERIAL_PORT )
+			{
+				PrintAndLog("Reconnect failed, retrying...");
+
+				if( txcmd_pending ) {
+					PrintAndLog("Cannot send bytes to offline proxmark");
+					txcmd_pending = false;
+				}
+
+				sleep(2);
 				continue;
 			}
-			UsbCommandReceived((UsbCommand*)rx);
+
+			PrintAndLog("Proxmark reconnected!");
+			need_reconnect = false;
+			offline = 0;
+			//CmdVersionW(NULL, true);
+			clearCommandBuffer();
+			uart_send(sp, (byte_t*) &version_cmd, sizeof(UsbCommand)); // request it from the HW
```

Comment:
This would destroy all data in  BigBuf[] when reconnecting. I think the major purpose of  the possibility to reconnect is to support offline functionality like sniffing some data?

Reply:
As it is now, unplugging causes the CPU to shoot to 100% used and the only way to continue what you're doing is to exit and restart the client.  When I'm on my laptop I sometimes need to move USB ports or occasionally it accidentally comes unplugged, and needing to restart is annoying.

At any rate I've seen some odd segfaults in readline with this, so I was thinking about ripping out the "re-get version info" stuff anyway.

Reply:
Also, as far as I can tell clearing the command buf doesn't touch BigBuf[].  Not sure if the act of disconnecting/reconnecting the USB cable will do that, though without a battery whatever's in the hardware will be gone.

Reply:
Newest version still has clearCommandBuffer() but instead of querying the version on re-connect it simply clears CmdVersion()'s cache.